akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMap.scala

```diff
@@ -174,8 +303,123 @@ final class ORMap[A, B <: ReplicatedData] private[akka] (
         case (None, None) ⇒ throw new IllegalStateException(s"missing value for $key")
       }
     }
+    val removedKeysIterator = (tombstoneVals.keys ++ that.tombstoneVals.keys).iterator
+    var removedValues = Map.empty[A, B]
+    removedKeysIterator.foreach { key ⇒
+      (this.tombstoneVals.get(key), that.tombstoneVals.get(key)) match {
+        case (Some(thisValue), Some(thatValue)) ⇒
+          if (thisValue.getClass != thatValue.getClass) {
+            val errMsg = s"Wrong type for merging [$key] in [${getClass.getName}], existing type " +
+              s"[${thisValue.getClass.getName}], got [${thatValue.getClass.getName}]"
+            throw new IllegalArgumentException(errMsg)
+          }
+          val removedValue = thisValue.merge(thatValue.asInstanceOf[thisValue.T]).asInstanceOf[B]
+          removedValues = removedValues.updated(key, removedValue)
+        case (Some(thisValue), None) ⇒
+          removedValues = removedValues.updated(key, thisValue)
+        case (None, Some(thatValue)) ⇒
+          removedValues = removedValues.updated(key, thatValue)
+        case (None, None) ⇒ throw new IllegalStateException(s"missing value for $key")
+      }
+    }
+    // tombstone the values for removed keys
+    new ORMap(mergedKeys, mergedValues, removedValues -- mergedKeys.elementsMap.keysIterator)
+  }
+
+  override def resetDelta: ORMap[A, B] =
+    if (delta.isEmpty) this
+    else new ORMap[A, B](keys, values, tombstoneVals)
+
+  override def mergeDelta(thatDelta: ORMap.DeltaOp): ORMap[A, B] = {
+    // helper function to simplify folds below
+    def foldValues(values: List[(A, ReplicatedData)], initial: B) =
+      values.foldLeft(initial) {
+        (acc: B, rhs: (A, ReplicatedData)) ⇒
+          {
+            val (_, value) = rhs
+            if (acc.isInstanceOf[DeltaReplicatedData] && value.isInstanceOf[ReplicatedDelta]) {
+              val accD = acc.asInstanceOf[DeltaReplicatedData]
+              accD.mergeDelta(value.asInstanceOf[accD.D]).asInstanceOf[B]
+            } else {
+              acc.merge(value.asInstanceOf[acc.T]).asInstanceOf[B]
+            }
+          }
+      }
+
+    val thatKeys: ORSet[A] = thatDelta match {
+      case d: AtomicDeltaOp[A, B] ⇒ keys.mergeDelta(d.underlying)
+      case ORMap.DeltaGroup(ops) ⇒
+        ops.foldLeft(keys)((acc, op) ⇒ acc.mergeDelta(op.asInstanceOf[AtomicDeltaOp[A, B]].underlying))
+    }
+
+    var mergedValues = Map.empty[A, B]
+    val mergedKeys = thatKeys
+    var thatValueDeltas: Map[A, List[(A, ReplicatedData)]] = Map.empty
+    var invalidatedTombstones: Map[A, Set[UniqueAddress]] = Map.empty
+
+    val processDelta: PartialFunction[ORMap.DeltaOp, Unit] = {
```

Comment:
both `putOp` and `removeOp` will invalidate the tombstone for the node from which they are coming from. This means that further `updated` deltas from that node will not be merged with tombstoned state. This is because in that light cone the previous state got destroyed either by `put` or `remove`.